enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1012 @@
+since OVN-Kubernetes (the component setting the finalizer) does not know when
+to remove the finalizer (i.e. it doesn't know the VMI lifecycle).
+
+This alternative also requires KubeVirt to template the pod indicating the name
+of the `IPAMClaim`, as described in the [pod templating](#pod-templating)
+section.
+
+### Delegate the IPAM functionality to a CNI IPAM plugin
+
+Below you can find a Network Attachment Definition requesting delegated IPAM
+from a separate IPAM plugin.
+
+```yaml
+---
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: whereabouts-conf
+spec:
+  config: '{
+      "cniVersion": "0.3.1",
+      "name": "l2-network",
+      "type": "ovn-k8s-cni-overlay",
+      "topology":"layer2",
+      "mtu": 1300,
+      "netAttachDefName": "default/whereabouts-conf",
+      "ipam": {
+        "type": "whereabouts",
+        "log_file": "/var/log/wb.log",
+        "log_level": "debug",
+        "range": "192.168.2.0/24"
+      }
+    }'
+```
+
+Thus, we would have the following three behaviors:
+- cni conf without a subnet defined **and** without IPAM key defined => only
+  static IP can be requested
+- cni conf without a subnet defined **but with** the IPAM key defined =>
+  delegate IPAM to an external plugin
+- cni conf with the subnet key defined => OVN-K provided IPAM for the provided
+  subnet
+
+**NOTE**: it results in an error if the user attempts to define **both** the
+subnet **and** the IPAM key at the same time.
+
+#### Why this alternative was not considered further
+
+Implementing the sticky IPs feature in separate IPAM plugins will prevent this
+solution from being used on the default cluster network for the HyperShift
+use cases, in which point to point routing is used to ensure the IPs persist
+during the migration of KubeVirt nodes. In fact, it will also not be usable for
+`layer3` topologies (secondary networks) since those topologies feature a
+subnet per node.
+
+Furthermore, performance wise, implementing the "sticky IPs" feature in the
+IPAM plugin - e.g. whereabouts - might not be a good option since we would need
+to invoke the IPAM plugin via CNI exec based API - i.e. invoke the CNI IPAM
+binary on disk, get the result (the IP), and then use said IP in the entire
+flow. Invoking the IPAM plugin is a synchronous operation that would block the
+OVN-Kubernetes plugin until it gets the response back.
+
+Adding insult to injury, given whereabouts instances actively compete for a
+datastore lock to update the IP pool allocations, need to check for the
+presence of overlapping IPs, and would - on top - have to also query the
+datastore for the presence of existing persistent allocations / persist new
+allocations, it is probable the time it takes to provision a pod to grow.
+
+Implementing a new CNI IPAM plugin - or refactoring whereabouts to address its
+shortcomings - would be a major effort.
+
+### Separate IPAMClaims controller
```

Comment:
So when I read the main proposal, I like it for what it means for OVN-K: it will give us a chance to improve its IPAM and its routing capabilities, hopefully along with a refactor improving its maintainability as well. 

But I don't get why from the kubevirt perspective the main proposal is better than this alternative. Why would kubevirt want to depend on OVN-K for this functionality? If this were to be handled through static IP requests, basically this alternative would potentially work with any CNI supporting such static IP requests which are already available in the multinetwork spec as well as probably many other IPAM implementations.

Reply:
I see, maybe the static IP request is not the best choice specifically. Still, the stance from kubevirt is it is ok for this functionality to depend on CNI plugins that support the new IPAMClaim specification, initially only being OVN-K? I am just curious if from kubevirt perspective other more modular approaches, that would potentially lower the cost of entry for CNIs to support, have been considered.